enterprise/frontend/src/metabase-enterprise/advanced_permissions/graph.ts

```diff
@@ -50,18 +67,32 @@ export function upgradeViewPermissionsIfNeeded(
     DataPermission.VIEW_DATA,
   );
 
-  if (
+  const requiresUnrestrictedAccess =
     value === DataPermissionValue.QUERY_BUILDER_AND_NATIVE &&
-    dbPermission !== DataPermissionValue.IMPERSONATED
-  ) {
+    ![
+      DataPermissionValue.UNRESTRICTED,
+      DataPermissionValue.IMPERSONATED,
+    ].includes(dbPermission);
+
+  const isGrantingQueryAccessWithBlockedChild =
+    value !== DataPermissionValue.NO &&
+    hasPermissionValueInSubgraph(
+      permissions,
+      groupId,
+      entityId,
+      database,
+      DataPermission.VIEW_DATA,
+      DataPermissionValue.BLOCKED,
+    );
+
+  if (requiresUnrestrictedAccess || isGrantingQueryAccessWithBlockedChild) {
     permissions = updateSchemasPermission(
       permissions,
       groupId,
       entityId,
       DataPermissionValue.UNRESTRICTED,
       database,
       DataPermission.VIEW_DATA,
-      false,
```

Comment:
this was the only place downgradeNative was being set to false and it's not a case we would have trigged it anyways. removed the option to control this since it never makes sense to leave it up to the caller to decide. (you'll see main refactor of this in `data-permissions.ts`